libraries/ESP8266WiFi/src/WiFiClientBearSSL.cpp

```diff
@@ -0,0 +1,1203 @@
+    case BR_ERR_X509_DN_MISMATCH: t = PSTR("Issuer/Subject DN mismatch in the chain."); break;
+    case BR_ERR_X509_BAD_SERVER_NAME: t = PSTR("Expected server name was not found in the chain."); break;
+    case BR_ERR_X509_CRITICAL_EXTENSION: t = PSTR("Unknown critical extension in certificate."); break;
+    case BR_ERR_X509_NOT_CA: t = PSTR("Not a CA, or path length constraint violation."); break;
+    case BR_ERR_X509_FORBIDDEN_KEY_USAGE: t = PSTR("Key Usage extension prohibits intended usage."); break;
+    case BR_ERR_X509_WEAK_PUBLIC_KEY: t = PSTR("Public key found in certificate is too small."); break;
+    case BR_ERR_X509_NOT_TRUSTED: t = PSTR("Chain could not be linked to a trust anchor."); break;
+  }
+  if (dest) { strncpy_P(dest, t, len); dest[len-1] = 0; }
+  return err;
+}
+
+bool WiFiClientBearSSL::probeMaxFragmentLength(const char* name, uint16_t port, uint16_t len)
+{
+  IPAddress remote_addr;
+  if (!WiFi.hostByName(name, remote_addr)) {
+    return false;
+  }
+  return WiFiClientBearSSL::probeMaxFragmentLength(remote_addr, port, len);
+}
+
+bool WiFiClientBearSSL::probeMaxFragmentLength(const String host, uint16_t port, uint16_t len)
+{
+  return WiFiClientBearSSL::probeMaxFragmentLength(host.c_str(), port, len);
+}
+
+// Checks for support of Maximum Frame Length Negotiation at the given
+// blocksize.  Note that, per spec, only 512, 1024, 2048, and 4096 are
+// supported.  Many servers today do not support this negotiation.
+
+// TODO - Allow for fragmentation...but not very critical as the ServerHello
+//      we use comes to < 80 bytes which has no reason to ever be fragmented.
+// TODO - Check the type of returned extensions and that the MFL is the exact
+//      same one we sent.  Not critical as only horribly broken servers would
+//      return changed or add their own extensions.
+bool WiFiClientBearSSL::probeMaxFragmentLength(IPAddress ip, uint16_t port, uint16_t len)
+{
+  // Hardcoded TLS 1.2 packets used throughout
+  static const uint8_t clientHelloHead_P[] PROGMEM = {
+    0x16, 0x03, 0x03, 0x00, 0, // TLS header, change last 2 bytes to len
+    0x01, 0x00, 0x00, 0, // Last 3 bytes == length
+    0x03, 0x03, // Proto version TLS 1.2
+    0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08, // Random (gmtime + rand[28])
+      0x11, 0x12, 0x13, 0x14, 0x15, 0x16, 0x17, 0x18,
+      0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08,
+      0x11, 0x12, 0x13, 0x14, 0x15, 0x16, 0x17, 0x18,
+    0x00, // Session ID
+  };
+  // Followed by our cipher-suite, generated on-the-fly
+  //    0x00, 0x02, // cipher suite len
+  //      0xc0, 0x13, // BR_TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA
+  static const uint8_t clientHelloTail_P[] PROGMEM = {
+    0x01, 0x00, // No compression
+    0x00, 0x05, // Extension length
+      0x00, 0x01, // Max Frag Len
+        0x00, 0x01, // len of MaxFragLen
+  };
+  // Followed by a 1-byte MFLN size requesst
+  //          0x04 // 2^12 = 4K
+  uint8_t mfl;
+
+  switch (len) {
```

Comment:
no default case

Reply:
It's there at line 1015 (returns false because MFLN protocol only lets you do very specific sizes).